### DISCUSSION TITLE
Contract with Fixed Point and non-Fixed Point ints

### DISCUSSION THREAD

**Discussion by joaomontenegro** [PaulRBerg/prb-math#43] [Q&A] [answered]
How to approach a contract that has some uint256 member variables as fixed point (via PRBMathUD60x18) and some other variables that non-fixed point uint256 integers (via, for example, SafeMath)?

**Comment by PaulRBerg** [comment, marked as answer]
~~First way is to use the [PRBMathSD59x18Typed.sol](https://github.com/hifi-finance/prb-math/tree/5d26ad9c2ff72e9ccd01fe11258e6f16664c3931#prbmathsd59x18typedsol) and [PRBMathUD60x18Typed.sol](https://github.com/hifi-finance/prb-math/tree/5d26ad9c2ff72e9ccd01fe11258e6f16664c3931#prbmathud60x18typedsol) flavours of PRBMath.~~

~~A second way, if you intend to use the vanilla flavours of the library, is to group the fixed-point types within structs, so that I know that if I ever use that struct, all of its containing integers are fixed-point numbers. E.g. [here](https://github.com/hifi-finance/hifi/blob/c6e585497e1e8a56aaeba71daad0dd4f08e5c8ae/packages/protocol/contracts/core/fintroller/SFintrollerV1.sol#L22-L27).~~

~~A third way, mentioned by @definim in https://github.com/hifi-finance/prb-math/discussions/29, is to use a different nomenclature for the fixed-point variables. E.g. he used Hungarian notation.~~

**Update** for PRBMath [v3.0.0](https://github.com/paulrberg/prb-math/releases/tag/v3.0.0):

Type safety is now offered via [user defined value types](udvt) `SD59x18` and `UD60x18`. You differentiate fixed-point from non-fixed-point types by simply declaring your variable with the value type (in fact, you must do this in V3, there is no way around it, as I have explained in this separate discussion https://github.com/paulrberg/prb-math/discussions/143).

**Reply by PaulRBerg** [reply]
Hey @joaomontenegro, I have recently released PRBMath V3, which is exactly what you were looking for when you posted this question in August 2021 :)

**Reply by joaomontenegro** [reply]
Awesome! Thank you!
